include/ua_server.h

```diff
@@ -1250,6 +1250,114 @@ UA_Server_triggerEvent(UA_Server *server, const UA_NodeId eventNodeId, const UA_
 
 #endif /* UA_ENABLE_SUBSCRIPTIONS_EVENTS */
 
+#ifdef UA_ENABLE_SUBSCRIPTIONS_ALARMS_CONDITIONS
+typedef enum UA_TwoStateVariableCallbackType {
+  UA_ENTERING_ENABLEDSTATE,
+  UA_ENTERING_ACKEDSTATE,
+  UA_ENTERING_CONFIRMEDSTATE,
+  UA_ENTERING_ACTIVESTATE
+} UA_TwoStateVariableCallbackType;
+
+/**
+ * callback prototype to set user specific callbacks
+ */
+typedef UA_StatusCode
+(*UA_TwoStateVariableChangeCallback)(UA_Server *server, const UA_NodeId *condition);
+
+/**
+ * create condition instance. The function checks first whether the passed conditionType
+ * is a subType of ConditionType. Then checks whether the condition source has HasEventSource
+ * reference to its parent. If not, a HasEventSource reference will be created between condition
+ * source and server object. To expose the condition in address space, a hierarchical ReferenceType
+ * should be passed to create the reference to condition source. Otherwise, UA_NODEID_NULL should be
+ * passed to make the condition unexposed.
+ * @param server The server object
+ * @param conditionType The NodeId of the node representation of the ConditionType
+ * @param conditionName The name of the condition to be created
+ * @param conditionSource The NodeId of the Condition Source (Parent of the Condition)
+ * @param hierarchialReferenceType The NodeId of Hierarchical ReferenceType between Condition and its source
+ * @param outNodeId The NodeId of the created Condition
+ * @return The StatusCode of the UA_Server_triggerEvent method */
+UA_StatusCode UA_EXPORT
+UA_Server_createCondition(UA_Server *server, const UA_NodeId conditionType,
+                          UA_QualifiedName conditionName, const UA_NodeId conditionSource,
+                          const UA_NodeId hierarchialReferenceType, UA_NodeId *outNodeId);
+
+/**
+ * set the value of condition field.
+ * @param server The server object
+ * @param condition The NodeId of the node representation of the Condition Instance
+ * @param variantValue Variant Value to be written to the Field
+ * @param type Variant Type.
+ * @param fieldName Name of the Field in which the value should be written
+ * @return The StatusCode of the UA_Server_triggerEvent method*/
+UA_StatusCode UA_EXPORT
+UA_Server_setConditionField(UA_Server *server, const UA_NodeId *condition,
+                            const void* variantValue, UA_UInt16 type,
+                            const UA_QualifiedName* fieldName);
+
+/**
+ * set the value of property of condition field.
+ * @param server The server object
+ * @param condition The NodeId of the node representation of the Condition Instance
+ * @param variantValue Variant Value to be written to the Field
+ * @param type Variant Type
+ * @param variableFieldName Name of the Field which has a property
+ * @param variablePropertyName Name of the Field Property in which the value should be written
+ * @return The StatusCode of the UA_Server_triggerEvent method*/
+UA_StatusCode UA_EXPORT
```

Comment:
Can I do all of this "manually" by browsing and writing in the information model?
This is not a critique. It would be rather good if we could implement this function with "public server API".

The comments above for pointers and variants applied as well.

Reply:
yes, it is possible, because those functions (UA_Server_setConditionField() and UA_Server_setConditionVariableFieldProperty()) use only public server API functions (UA_Server_browseSimplifiedBrowsePath(), UA_Variant_setScalar() and UA_Server_writeValue()).


> This is not a critique. It would be rather good if we could implement this function with "public server API".
I am afraid I didn't understand that part of the comment:(